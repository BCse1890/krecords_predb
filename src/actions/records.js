import uuid from 'uuid';
import database from '../firebase/firebase';
import moment from 'moment';

// ADD RECORD
export const addRecord = (record) => ({
    type: 'ADD_RECORD',
    record
});

export const startAddRecord = (recordData = {}) => {
    return (dispatch) => {
        const {
            name='', 
            address='', 
            contactTel='', 
            email='', 
            relativeName='',
            relativeContactTel='', 
            mainDojo='', 
            dateJoinClub=0, 
            gradeLevel=10, 
            gradingDate=undefined, 
            classesSinceGrading=0
        } = recordData;
        
        const record = {name, mainDojo, gradeLevel, gradingDate, dateJoinClub, 
            address, contactTel, email, relativeName, relativeContactTel, 
            classesSinceGrading };

        database.ref('records').push(record).then((ref) => {
            dispatch(addRecord({
                id: ref.key,
                ...record
            }));
        });
    };
};

/* 
// ADD RECORD
export const addRecord = (record) => ({
    type: 'ADD_RECORD',
    record
});

export const startAddRecord = (recordData = {}) => {
    return (dispatch) =>  {
        const {
            name='', 
            address='', 
            contactTel='', 
            email='', 
            relativeName='',
            relativeContactTel='', 
            mainDojo='', 
            dateJoinClub=0, 
            gradeLevel=10, 
            gradingDate=0, 
            classesSinceGrading=0
        } = recordData;
        
        const record = {name, mainDojo, gradeLevel, gradingDate, dateJoinClub, 
            address, contactTel, email, relativeName, relativeContactTel, 
            classesSinceGrading };

        // return next line so that can use chained promises in test
        // records.test.js    
        database.ref('records').push(record).then((ref) => {
            dispatch(addRecord({
                id: ref.key,
                ...record
            }));
        });
    };
};
 */
// DELETE RECORD
export const deleteRecord = ({ id } = {}) => ({
    type: 'DELETE_RECORD',
    id
});

// EDIT RECORD
export const editRecord = (id, updates) => ({
    type: 'EDIT_RECORD',
    id,
    updates
});