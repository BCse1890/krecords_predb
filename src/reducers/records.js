
// create kRecords Reducer
const recordsReducerDefault = [];

const recordsReducer = (state = recordsReducerDefault, action) => {
    switch(action.type) {
        case 'ADD_RECORD':
            return [
                //spreads out all existing records in array
                // and adds on a new array through action.record
                ...state,
                action.record
            ];   
        case 'DELETE_RECORD':
            return state.filter(({ id }) => (id !== action.id) );
        case 'EDIT_RECORD':
            return state.map((record) => {
                    if(record.id === action.id) {
                        return {
                            ...record,
                            ...action.updates
                        }
                    } else {
                        return record;
                    }
                });
        default:
            return state;
    }
};

export default recordsReducer;