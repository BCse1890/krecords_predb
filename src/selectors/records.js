import moment from 'moment';

// Get Visible Expenses
export default (records, { text, sortBy, startDate, endDate }) => {
//const getVisibleExpenses = (records, { text, gradingDate, sortBy, startDate, endDate }) => {
    return records.filter((record) => {
        // filter out if startDate not a number or record.created >= startDate;
        const gradeDateMoment = moment(record.gradingDate);
        const startDateMatch = startDate ? startDate.isSameOrBefore(gradeDateMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(gradeDateMoment, 'day') : true;
        
        const textNameMatch = record.name.toLowerCase().includes(text.toLowerCase());
        const textDojoMatch = record.mainDojo.toLowerCase().includes(text.toLowerCase());
        
            
        return startDateMatch && endDateMatch && textDojoMatch || textNameMatch;
    }).sort((a, b) => {
        if(sortBy === 'mainDojo') {
            return a.mainDojo < b.mainDojo ? -1 : 1;
        } else if(sortBy === 'gradeLevel') {
            return a.gradeLevel < b.gradeLevel ? 1 : -1;
        }
    });
};

// export default getVisibleExpenses;