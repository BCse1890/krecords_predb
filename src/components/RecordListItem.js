import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { deleteRecord } from '../actions/records';


const RecordListItem = ({ id, name, mainDojo, gradeLevel, gradingDate, classesSinceGrading }) => (
    <div>
        <Link to={`/edit/${id}`}>
          <h3>Name: {name}</h3>
        </Link>
        <p>Main dojo: {mainDojo}</p>
        <p>Grade level: {gradeLevel}</p>
        <p>Grading date: {gradingDate}</p>
        <p>Classes since grading: {classesSinceGrading}</p>
        <Link to={`/contact/${id}`}>
          <h3>Registration and Contact details</h3>
        </Link>
        <br/>
    </div>
);


export default RecordListItem;