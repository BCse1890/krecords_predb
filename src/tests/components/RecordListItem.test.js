import React from 'react';
import { shallow, mount, render } from 'enzyme';
import '../setUpTests';
import records from '../fixtures/records';
import RecordListItem from '../../components/RecordListItem';

test('should render RecordListItem', () => {
    const wrapper = shallow(<RecordListItem {...records[1]} />);
    expect(wrapper).toMatchSnapshot();
});