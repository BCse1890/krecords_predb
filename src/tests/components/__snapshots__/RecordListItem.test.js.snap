// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RecordListItem 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/2"
  >
    <h3>
      Name: 
      Bo Ko
    </h3>
  </Link>
  <p>
    Main dojo: 
    Emersons Green
  </p>
  <p>
    Grade level: 
    4
  </p>
  <p>
    Grading date: 
    2
  </p>
  <p>
    Classes since grading: 
    0
  </p>
  <Link
    replace={false}
    to="/contact/2"
  >
    <h3>
      Registration and Contact details
    </h3>
  </Link>
  <br />
</div>
`;
